onnxruntime/python/tools/kernel_explorer/kernels/groupnorm_test.py

```diff
@@ -0,0 +1,214 @@
+
+    num_groups = [32]
+    return product(batch_sizes, height, num_channels, num_groups)
+
+
+def dtype_to_funcs(dtype):
+    type_map = {
+        "float16": list(filter(lambda x: re.search("GroupNormNHWC.*_half", x), dir(ke))),
+        "float32": list(filter(lambda x: re.search("GroupNormNHWC.*_float", x), dir(ke))),
+    }
+    return type_map[dtype]
+
+
+def sigmoid_function(x):
+    return 1.0 / (1.0 + np.exp(-x))
+
+
+def group_norm(input_x, gamma, beta, num_groups, epsilon, with_swish):
+    n, h, w, c = input_x.shape
+    input_x = input_x.transpose([0, 3, 1, 2])
+    assert c % num_groups == 0
+    x = input_x.reshape((n, num_groups, -1))
+    mean = np.mean(x, axis=-1, keepdims=True)
+    var = np.var(x, axis=-1, keepdims=True)
+    x = (x - mean) / np.sqrt(var + epsilon)
+    x = x.reshape((n, c, h, w))
+    x = x.transpose([0, 2, 3, 1])
+    x = x * gamma + beta
+
+    if with_swish:
+        x = x * sigmoid_function(x)
+    return x
+
+
+def run_group_norm(batch_size: int, height: int, num_channels: int, num_groups: int, dtype: str, func):
+    np.random.seed(0)
+    width = height
+    input_x = np.random.rand(batch_size, height, width, num_channels).astype(np.float32)
+    gamma = np.random.rand(num_channels).astype(np.float32)
+    beta = np.random.rand(num_channels).astype(np.float32)
+    # the size of workspace is defined in onnxruntime/contrib_ops/cuda/diffusion/group_norm_impl.h L18
+    workspace = np.random.rand((np.dtype(np.float32).itemsize * 2) * 32 * 32).astype(np.float32)
+    epsilon = 1e-05
+    output_y = np.random.rand(batch_size, height, width, num_channels).astype(dtype)
+    use_swish = True
+
+    host_x = input_x.astype(dtype)
+    input_d = ke.DeviceArray(host_x)
+    gamma_d = ke.DeviceArray(gamma)
+    beta_d = ke.DeviceArray(beta)
+    workspace_d = ke.DeviceArray(workspace)
+    y_d = ke.DeviceArray(output_y)
+    f = getattr(ke, func)
+
+    my_op = f(
+        y_d,
+        workspace_d,
+        input_d,
+        gamma_d,
+        beta_d,
+        batch_size,
+        height,
+        width,
+        num_channels,
+        num_groups,
+        epsilon,
+        use_swish,
+    )
+    if my_op.IsSupported():
+        my_op.Run()
+
+        y_d.UpdateHostNumpyArray()
+
+        y_ref = group_norm(input_x, gamma, beta, num_groups, epsilon, use_swish).astype(dtype)
+        np.testing.assert_allclose(y_ref, output_y, atol=1e-02)
+
+
+dtypes = ["float32", "float16"]
+
+
+@pytest.mark.parametrize("sd_sizes", get_sd_sizes())
+@pytest.mark.parametrize("dtype", dtypes)
+def test_skip_layer_norm(sd_sizes, dtype):
+    for func in dtype_to_funcs(dtype):
+        run_group_norm(*sd_sizes, dtype, func)
+
+
+@dataclass
+class GroupNormNHWCMetric(ke.BandwidthMetric):
+    batch_size: int
+    height: int
+    width: int
+    num_channels: int
+
+    def report(self):
+        prefix = f"{self.name:<50} {self.dtype}  height={self.height:<4} width={self.width:<4} num_channels={self.num_channels:<4} "
```

Comment:
I stumbled that we should put the name at last column, because it is generally unexpected long and uneven after CK is added, and it makes the command line log unreadable...

Reply:
Done.